File: protocol-core/src/main/kotlin/com/airninja/protocol/transport/SecureChannel.kt
package com.airninja.protocol.transport

import com.airninja.protocol.identity.DeviceIdentity
import com.airninja.protocol.pairing.Sas
import com.airninja.protocol.secure.NoiseHandshake
import com.airninja.protocol.secure.NoiseRole
import com.airninja.protocol.secure.TransportPair
import java.io.InputStream
import java.io.OutputStream

class SecureChannel internal constructor(
    private val transport: TransportPair,
    val handshakeHash: ByteArray,
    val remoteStaticPublicKey: ByteArray,
    private val input: InputStream,
    private val output: OutputStream,
) {
    val sas: String
        get() = Sas.derive(handshakeHash)

    fun send(frame: ByteArray) {
        StreamFraming.writeFrame(output, transport.encrypt(frame))
    }

    fun receive(): ByteArray = transport.decrypt(StreamFraming.readFrame(input))

    companion object {
        fun handshake(
            role: NoiseRole,
            identity: DeviceIdentity,
            input: InputStream,
            output: OutputStream,
        ): SecureChannel {
            val handshake = NoiseHandshake.create(role, identity)
            handshake.start()
            while (!handshake.isHandshakeFinished) {
                if (handshake.needsWrite()) {
                    StreamFraming.writeFrame(output, handshake.writeMessage())
                } else {
                    handshake.readMessage(StreamFraming.readFrame(input))
                }
            }
            val handshakeHash = handshake.handshakeHash
            val remoteStaticPublicKey = handshake.remoteStaticPublicKey
            return SecureChannel(handshake.split(), handshakeHash, remoteStaticPublicKey, input, output)
        }
    }
}

File: macos/AirNinjaCore/Sources/AirNinjaCore/Transport/SecureChannel.swift
import Foundation

public final class SecureChannel {
    private let transport: NoiseTransport
    private let stream: ByteStream

    public let handshakeHash: Data
    public let remoteStaticPublicKey: Data

    public var sas: String { Sas.derive(handshakeHash: handshakeHash) }

    init(transport: NoiseTransport, handshakeHash: Data, remoteStaticPublicKey: Data, stream: ByteStream) {
        self.transport = transport
        self.handshakeHash = handshakeHash
        self.remoteStaticPublicKey = remoteStaticPublicKey
        self.stream = stream
    }

    public func send(_ frame: Data) throws {
        try StreamFraming.writeFrame(stream, try transport.encrypt(frame))
    }

    public func receive() throws -> Data {
        try transport.decrypt(try StreamFraming.readFrame(stream))
    }

    public static func handshake(role: NoiseRole, identity: DeviceIdentity, stream: ByteStream) throws -> SecureChannel {
        let handshake = try NoiseHandshake(role: role, identity: identity)
        while !handshake.isComplete {
            if handshake.needsWrite() {
                try StreamFraming.writeFrame(stream, try handshake.writeMessage())
            } else {
                _ = try handshake.readMessage(try StreamFraming.readFrame(stream))
            }
        }
        guard let remoteStaticPublicKey = handshake.remoteStaticPublicKey else {
            throw NoiseError.handshakeIncomplete
        }
        return SecureChannel(
            transport: try handshake.split(),
            handshakeHash: handshake.handshakeHash,
            remoteStaticPublicKey: remoteStaticPublicKey,
            stream: stream
        )
    }
}
